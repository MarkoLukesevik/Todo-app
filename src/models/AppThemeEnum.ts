export enum AppThemeEnum {
    Light = 0,
    Dark = 1
}
