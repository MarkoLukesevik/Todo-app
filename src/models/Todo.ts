export interface Todo {
  id: string
  title: string
  isComplete: false
}
