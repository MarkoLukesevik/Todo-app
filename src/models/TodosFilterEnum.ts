export enum TodosFilterEnum {
    All = 0,
    Active = 1,
    Completed = 2
}