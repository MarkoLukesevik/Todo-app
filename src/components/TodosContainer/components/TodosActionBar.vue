<script setup lang="ts">
import { computed } from 'vue';

import { AppThemeEnum } from '@/models/AppThemeEnum';
import { TodosFilterEnum } from '@/models/TodosFilterEnum';

const props = defineProps<{
    appTheme: AppThemeEnum;
    todosFilter: TodosFilterEnum;
}>();

const emit = defineEmits<{
    (e: 'change-todos-filter', value: TodosFilterEnum): void;
}>();

const actionBarClasses = computed<string[]>(() => {
    const classes = ['action-bar'];

    if (props.appTheme === AppThemeEnum.Dark)
        classes.push('action-bar__dark');

    return classes;
});
</script>

<template>
    <div :class="actionBarClasses">
        <span
            class="action-bar__filter"
            :class="{ 'action-bar__active-filter': todosFilter === TodosFilterEnum.All }"
            @click="emit('change-todos-filter', TodosFilterEnum.All)"
        >
            All
        </span>
        <span
            class="action-bar__filter"
            :class="{ 'action-bar__active-filter': todosFilter === TodosFilterEnum.Active }"
            @click="emit('change-todos-filter', TodosFilterEnum.Active)"
        >
            Active
        </span>
        <span
            class="action-bar__filter"
            :class="{ 'action-bar__active-filter': todosFilter === TodosFilterEnum.Completed }"
            @click="emit('change-todos-filter', TodosFilterEnum.Completed)"
        >
            Completed
        </span>
    </div>
</template>

<style scoped lang="scss">
.action-bar {
    display: flex;
    flex-direction: row;
    flex: 0 1 auto;
    align-items: center;
    justify-content: center;
    column-gap: 1rem;

    background-color: #ffffff;
    color: #9495A5;
    border-radius: 5px;
    padding: 1rem;

    font-family: 'Josefin Sans', sans-serif;
    font-size: 0.9rem;
    font-weight: 600;

    &__filter {
        cursor: pointer;
    }

    &__dark {
        background-color: #25273D;
        color: #5B5E7E;
    }

    & .action-bar__active-filter {
        color: #3A7CFD;
    }
}

@media only screen and (min-width: 768px) {
    .action-bar {
        display: none;
    }
}
</style>