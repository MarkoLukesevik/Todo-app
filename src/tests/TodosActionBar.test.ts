import { beforeEach, describe, expect, it } from "vitest";
import { shallowMount } from "@vue/test-utils";

import { AppThemeEnum } from "@/models/AppThemeEnum";
import { TodosFilterEnum } from "@/models/TodosFilterEnum";

import TodosActionBar from "@/components/TodosContainer/components/TodosActionBar.vue";

describe('Test suite for todos actions bar component', () => {
    let wrapper: any;

    beforeEach(() => {
        wrapper = shallowMount(TodosActionBar, {
            props: {
                appTheme: AppThemeEnum.Light,
                todosFilter: TodosFilterEnum.All
            }
        });
    });

    it('Should match the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it('Should apply correct classes based on appTheme prop', async () => {
        const actionbar = await wrapper.find('[data-test="action-bar"]');
        expect(actionbar.classes()).toEqual(['action-bar']);

        await wrapper.setProps({ appTheme: AppThemeEnum.Dark });

        expect(actionbar.classes()).toEqual(['action-bar', 'action-bar__dark']);
    });

    it('Should apply correct active filter class based on appTheme prop', async () => {
        const allFilter = await wrapper.find('[data-test="action-bar-all-filter"]');
        const activeFilter = await wrapper.find('[data-test="action-bar-active-filter"]');
        const completedFilter = await wrapper.find('[data-test="action-bar-completed-filter"]');

        expect(allFilter.classes()).toEqual(['action-bar__filter', 'action-bar__active-filter']);
        expect(activeFilter.classes()).toEqual(['action-bar__filter']);
        expect(completedFilter.classes()).toEqual(['action-bar__filter']);

        await wrapper.setProps({ todosFilter: TodosFilterEnum.Active});

        expect(allFilter.classes()).toEqual(['action-bar__filter']);
        expect(activeFilter.classes()).toEqual(['action-bar__filter', 'action-bar__active-filter']);
        expect(completedFilter.classes()).toEqual(['action-bar__filter']);

        await wrapper.setProps({ todosFilter: TodosFilterEnum.Completed});

        expect(allFilter.classes()).toEqual(['action-bar__filter']);
        expect(activeFilter.classes()).toEqual(['action-bar__filter']);
        expect(completedFilter.classes()).toEqual(['action-bar__filter', 'action-bar__active-filter']);
    });

    it('Should emit the correct event with the filter value when a filter is clicked', async () => {
        const allFilter = await wrapper.find('[data-test="action-bar-all-filter"]');
        const activeFilter = await wrapper.find('[data-test="action-bar-active-filter"]');
        const completedFilter = await wrapper.find('[data-test="action-bar-completed-filter"]');

        allFilter.trigger('click');
        expect(wrapper.emitted('change-todos-filter')[0]).toEqual([TodosFilterEnum.All]);
    
        activeFilter.trigger('click');
        expect(wrapper.emitted('change-todos-filter')[1]).toEqual([TodosFilterEnum.Active]);
    
        completedFilter.trigger('click');
        expect(wrapper.emitted('change-todos-filter')[2]).toEqual([TodosFilterEnum.Completed]);
      });
});