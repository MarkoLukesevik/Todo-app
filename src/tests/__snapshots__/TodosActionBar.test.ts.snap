// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test suite for todos actions bar component > Should match the snapshot 1`] = `
<div
  class="action-bar"
  data-test="action-bar"
  data-v-ad5062d3=""
>
  <span
    class="action-bar__filter action-bar__active-filter"
    data-test="action-bar-all-filter"
    data-v-ad5062d3=""
  >
     All 
  </span>
  <span
    class="action-bar__filter"
    data-test="action-bar-active-filter"
    data-v-ad5062d3=""
  >
     Active 
  </span>
  <span
    class="action-bar__filter"
    data-test="action-bar-completed-filter"
    data-v-ad5062d3=""
  >
     Completed 
  </span>
</div>
`;
